### ISSUE TITLE
upgrading sysrepo and keeping the startup datastore around

### ISSUE THREAD

**Issue by jktjkt** [sysrepo/sysrepo#2280]
**TL;DR**: how should we handle system updates to make sure that sysrepo database stays "healthy"?

In c226ab112470fc4be7718f7c84f9604ddf49ff65 (November 2020), some fields in the internal `sysrepo-monitoring` module were renamed from `pid` to `cid` without a revision bump of that module. If a post-November sysrepo ever runs on a system with pre-November YANG module, sysrepo operations fail with this error:
```
Failed to find "cid" as a sibling to "sysrepo-monitoring:datastore".
```
...and we started hitting just that in our lab.

On our devices, we're using read-only [A/B filesystems for updates](https://rauc.readthedocs.io/en/latest/scenarios.html#symmetric-root-fs-slots). We also mount a RAM filesystem overlay at `/etc`, and there's a pair of A/B persistent data partitions (`/cfg`) whose content is applied at boot time, "overwriting" whatever was stored in the matching rootfs' read-only `/etc`. So our system always boots from a read-only partition (say, *A*), and when we upgrade the system to a newer version, the updater writes out a completely new system `rootfs` to the other half (say, *B*), and then the script somehow copies around the persistent configuration data from *cfg-A* to *cfg-B*. The whole idea behind this is that if the boot to the just-updated FW slot fails for some reason, we can easily revert to the previously working version (and we have this mostly automated, because the appliances will be located "in the middle of nowhere" along our DWDM network).

The `rootfs` image that we are installing is used across different device types which use different set of application-specific YANG modules (including different YANG features enabled for some of these modules). As such, none of these app-specific YANG modules are installed when building the image. Instead, during boot, the SW checks the HW and determines what device type it's running on, and installs the required YANG modules (+features). This means that we do not really require persistent storage of application-specific YANG modules, although we currently copy the whole `/etc/sysrepo` into `/cfg/etc/sysrepo` -- both the `data` and `yang` directories, which is probably a bug. (BTW, we're also invoking Netopeer's installation shell scripts during the first boot. That way, the SSH host key is not in the build image, and there are no issues with file permissions because our build system requires extra work for `root`-owned `0600` files.)

In practical terms, this means that at the time we build the firmware image, we always install sysrepo from scratch (there's no previous state), but once the target system boots up into the new software, the configuration stored in a persistent configuration partition override the newer version of `/etc/sysrepo/yang/sysrepo-monitoring@2020-04-17.yang`, and sysrepo encounters an older copy of that file which uses `pid` instead of `cid`.

I would like to ask for your help on how to best "fix this". What we are after is a way of preserving the system configuration (the contents of the `startup` datastore) across system updates. Today, we have an `inotify` FS watcher which simply copies around changed `*.startup` files, like this:
```bash
while true; do
  inotifywait -e CLOSE_WRITE /etc/sysrepo/data/*.startup \
    && mkdir -p /cfg/etc/sysrepo/data/ \
    && cp -a /etc/sysrepo/data/*.startup /cfg/etc/sysrepo/data/
done
```
...which makes sure that once an `<edit-config>` for the `startup` DS is committed, it is made persistent via the config partition. Is this safe, from your point of view?

Will sysrepo handle the following workflow:

1. no sysrepo-using applications are running,
2. `/etc/sysrepo` contains whatever gets installed during sysrepo's own `make install`, and **nothing else**,
3. all the `*.startup` files are copied around, including those for which there are no corresponding `*.yang` files around (yet),
4. netopeer2's config scripts get executed: `setup.sh`, `merge_hostkey.sh` and `merge_config.sh`,
5. all application-specific YANG files are installed, and YANG features are enabled, both via `sysrepoctl`,
6. if there's no matching `foo.startup` and the application-specific `foo` YANG module requires some initial config, it is applied via `sysrepocfg --import`,
7. `netopeer2-server` starts,
8. application-specific daemons start.

Is this a reasonable workflow? Should we do that in a different way?

I would actually prefer having the data in JSON or XML formats, but a naive `sysrepocfg -d startup -f json -i` of all modules at once won't work (it would clash with netopeer's installation scripts at least). Will sysrepo work if the `*.startup` files were JSON, for example? This will need extra work because `sysrepocfg -X` errors out for those modules that don't have any data leafs, but it would make it easier to implement application-specific schema updates during major YANG module revisions.

**Comment by jktjkt** [comment]
The steps outlined above do not appear to work. After step 3:
```console
# sysrepoctl --list
[ERR]: Data model "ietf-netconf-acm" not found.
sysrepoctl error: Failed to connect (libyang error)
```
I would appreciate your help here.

**Comment by michalvasko** [comment]
Firstly, as for renaming `pid` to `cid` in the module without revision bump, sorry about that. But I am currently working on some new features of *sysrepo* that include some changes to this module as well. I have already bumped the module version for that so once that is merged (a few days, I hope), it should be fixed.

Now, direct changes of *sysrepo* repository are never a good idea because they are simply not expected, which can lead to all kinds of problems. You should always use *sysrepocfg* when working with the configuration in any way, it should support any operation. Note that instead of `sysrepocfg -I=<file>` you can use `sysrepocfg -E=<file>` which essentially does configuration merge instead of replace.

**Comment by jktjkt** [comment]
I understand that you do not support changing sysrepo's database "behind its back" (and I agree with this), so let me rephrase. The problem that I'm facing ls (IMHO) equivalent to a backup + restore.

How do I backup the content of the `startup` datastore and restore it to another device? We can assume that that "another device" has just-installed sysrepo and just-installed-and-configured Netopeer2. I have the information about what YANG modules and their features available out-of-band. I prefer the backup format to be either XML or JSON, but I can live with anything, as long as it supports backup/recoveries across sysrepo versions starting today.

**Comment by michalvasko** [comment]
Yeah, I think I understand what you mean and I answered. You should use `sysrepocfg -X` to backup the datastore and `sysrepocfg -I` or `sysrepocfg -E` to restore the datastore. You said you cannot do this because `sysrepocfg -I` would overwrite any data of *netopeer2* and you are correct, that is why I suggested to use `sysrepocfg -E` to avoid this problem.

**Comment by jktjkt** [comment]
Thanks, I got it working. There's one thing which puzzled me -- once I install some YANG model and initialize its data like this:
```sh
sysrepoctl --search-dirs ${YANG_DIR} --install ${YANG_DIR}/czechlight-roadm-device@2019-09-30.yang
sysrepoctl --change czechlight-roadm-device --group optics --permissions 0664 --apply
sysrepoctl --change czechlight-roadm-device --enable-feature ${WITH_FEATURE}
sysrepocfg --datastore=startup --format=json --module=czechlight-roadm-device --import="${YANG_DIR}/${CZECHLIGHT}.json"
```
..and then I import the full `startup` DS (the JSON file also contains data for the `czechlight-roadm-device` model that I just initialized above):
```sh
sysrepocfg -d startup -f json --import=/cfg/sysrepo/startup.json
```
Then I start my application which subscribes to various subtrees of the `czechlight-roadm-device` module, but the app does not get the initial content of that module. Is that expected? If I run `sysrepocfg -C startup` after the `startup` DS got updated, but before launching my app, stuff appears to work correctly and the app gets the expected data via the initial callback (`SR_EV_ENABLED`). Is this expected, and is this documented?

**Comment by michalvasko** [comment]
Yes, it is expected (and documented, I mentioned it in https://github.com/CESNET/netopeer2/pull/802). Data are not copied from `startup` to `running` on their own in your case, you must do so explicitly.

**Comment by jktjkt** [comment]
Yup, @syyyr explained to me that I keep thinking about the legacy-sysrepo behavior, sorry.

In case someone comes looking for a solution in future, here's what we've got:

#### netopeer2-install-yang.service
```ini
[Unit]
Description=Install YANG for Netopeer2

[Service]
Type=oneshot
Environment=NP2_MODULE_DIR=/usr/share/yang/modules/netopeer2 NP2_MODULE_PERMS=0600 USER=root
ExecStart=/usr/libexec/netopeer2/setup.sh

[Install]
WantedBy=multi-user.target
```

#### cfg-restore-sysrepo.service 
```ini
[Unit]
Description=Restore sysrepo startup datastore from /cfg
After=netopeer2-install-yang.service czechlight-install-yang.service cfg.mount
Requires=netopeer2-install-yang.service czechlight-install-yang.service cfg.mount
Before=netopeer2-setup.service netopeer2.service sysrepo-persistent-cfg.service
ConditionPathExists=/cfg/sysrepo/startup.json

[Service]
Type=oneshot
ExecStart=/bin/sysrepocfg -d startup -f json --import=/cfg/sysrepo/startup.json
ExecStart=/bin/sysrepocfg -C startup

[Install]
WantedBy=multi-user.target
```

#### netopeer2-setup.service
```ini
[Unit]
Description=Install initial configuration for Netopeer2
Requires=netopeer2-install-yang.service
After=netopeer2-install-yang.service
Before=netopeer2.service

[Service]
Type=oneshot
ExecStart=/usr/libexec/netopeer2/ensure-config.sh

[Install]
WantedBy=multi-user.target
```
#### netopeer2.service 
```ini
[Unit]
Description=A NETCONF server on top of sysrepo
After=syslog.target network.target netopeer2-setup.service
Requires=netopeer2-setup.service
StartLimitIntervalSec=0

[Service]
Type=forking
ExecStart=/usr/bin/netopeer2-server
PIDFile=/var/run/netopeer2-server.pid
PrivateTmp=yes
PrivateDevices=yes
ProtectSystem=yes
ProtectHome=yes
SystemCallArchitectures=native
Restart=always

[Install]
WantedBy=multi-user.target
```
#### ensure-config.sh
This one only needs to be defined because of https://github.com/systemd/systemd/issues/18207; when this is fixed I can just use `ExecCondition=`.
```bash
#!/usr/bin/env bash

set -e

# optional env variable override
if [ -n "$SYSREPOCFG_EXECUTABLE" ]; then
    SYSREPOCFG="$SYSREPOCFG_EXECUTABLE"
# avoid problems with sudo PATH
elif [ `id -u` -eq 0 ]; then
    SYSREPOCFG=`su -c 'which sysrepocfg' -l $USER`
else
    SYSREPOCFG=`which sysrepocfg`
fi

SERVER_CONFIG=`$SYSREPOCFG -d startup -X -x "/ietf-netconf-server:netconf-server/listen/endpoint[1]/name | /ietf-netconf-server:netconf-server/call-home/netconf-client[1]/name"`
if [ -n "$SERVER_CONFIG" ]; then
    echo "ietf-netconf-server with listen or call-home already present"
    exit 0
fi

echo "Configuring default SSH listen address"
/usr/libexec/netopeer2/merge_hostkey.sh
echo "Generating the SSH host key"
/usr/libexec/netopeer2/merge_config.sh
```

#### sysrepo-persistent-cfg.service 
```ini
[Unit]
Description=Persisting persistent sysrepo datastores to /cfg
After=netopeer2.service cfg.mount
Requires=netopeer2.service cfg.mount

[Service]
Type=simple
UMask=0077
ExecStart=/bin/sh -c 'while true; do inotifywait -e CLOSE_WRITE /etc/sysrepo/data/*.startup && mkdir -p /cfg/sysrepo/ && sysrepocfg -d startup -f json -X > /cfg/sysrepo/startup.json; done'
```


The other referenced files (`setup.sh`, `merge_config.sh` and `merge_hostkey.sh`) come from Netopeer2. Here are [changes to the Buildroot recipe](https://gerrit.cesnet.cz/plugins/gitiles/github/buildroot/buildroot/+/refs/heads/cesnet/2021-01-11).

Closing; let's hope that this is magic enough to handle schema updates in future.
